distributed/tests/test_worker_memory.py

```diff
@@ -0,0 +1,680 @@
+async def test_override_data_worker(s):
+    # Use a UserDict to sidestep potential special case handling for dict
+    async with Worker(s.address, data=UserDict) as w:
+        assert type(w.data) is UserDict
+
+    data = UserDict({"x": 1})
+    async with Worker(s.address, data=data) as w:
+        assert w.data is data
+        assert w.data == {"x": 1}
+
+
+@gen_cluster(
+    client=True,
+    nthreads=[("", 1)],
+    Worker=Nanny,
+    worker_kwargs={"data": UserDict},
+)
+async def test_override_data_nanny(c, s, n):
+    r = await c.run(lambda dask_worker: type(dask_worker.data))
+    assert r[n.worker_address] is UserDict
+
+
+@gen_cluster(
+    client=True,
+    nthreads=[("", 1)],
+    worker_kwargs={"memory_limit": "1 GB", "data": UserDict},
+    config={"distributed.worker.memory.monitor-interval": "10ms"},
+)
+async def test_override_data_vs_memory_monitor(c, s, a):
+    a.monitor.get_process_memory = lambda: 801_000_000 if a.data else 0
+    assert memory_monitor_running(a)
+
+    # Push a key that would normally trip both the target and the spill thresholds
+    class C:
+        def __sizeof__(self):
+            return 801_000_000
+
+    # Capture output of log_errors()
+    with captured_logger(logging.getLogger("distributed.utils")) as logger:
+        x = c.submit(C)
+        await wait(x)
+
+        # The pause subsystem of the memory monitor has been tripped.
+        # The spill subsystem hasn't.
+        while a.status != Status.paused:
+            await asyncio.sleep(0.01)
+        await asyncio.sleep(0.05)
+
+    # This would happen if memory_monitor() tried to blindly call SpillBuffer.evict()
+    assert "Traceback" not in logger.getvalue()
+
+    assert type(a.data) is UserDict
+    assert a.data.keys() == {x.key}
+
+
+class ManualEvictDict(UserDict):
+    """A MutableMapping which implements distributed.spill.ManualEvictProto"""
+
+    def __init__(self):
+        super().__init__()
+        self.evicted = set()
+
+    @property
+    def fast(self):
+        # Any Sized of bool will do
+        return self.keys() - self.evicted
+
+    def evict(self):
+        # Evict a random key
+        k = next(iter(self.fast))
+        self.evicted.add(k)
+        return 1
+
+
+@gen_cluster(
+    client=True,
+    nthreads=[("", 1)],
+    worker_kwargs={"memory_limit": "1 GB", "data": ManualEvictDict},
+    config={
+        "distributed.worker.memory.pause": False,
+        "distributed.worker.memory.monitor-interval": "10ms",
+    },
+)
+async def test_manual_evict_proto(c, s, a):
+    """data is a third-party dict-like which respects the ManualEvictProto duck-type
+    API. spill threshold is respected.
+    """
+    a.monitor.get_process_memory = lambda: 701_000_000 if a.data else 0
+    assert memory_monitor_running(a)
+    assert isinstance(a.data, ManualEvictDict)
+
+    futures = await c.scatter({"x": None, "y": None, "z": None})
+    while a.data.evicted != {"x", "y", "z"}:
+        await asyncio.sleep(0.01)
```

Comment:
For my own comprehension, this test mocks the process memory to above the 70% limit, which forces the "x", "y" and "z" keys to be evicted as per the `ManualEvictDict` protocol.